pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,448 @@
+			let asset_id: T::AssetId =
+				AssetTypeId::<T>::get(&asset_type).ok_or(Error::<T>::AssetTypeDoesNotExist)?;
+			AssetMetadatas::<T>::insert(&asset_id, &metadata);
+
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+			Ok(())
+		}
+
+		/// Change the amount of units we are charging per execution second
+		/// for a given ForeignAssetType
+		#[pallet::weight(T::WeightInfo::set_asset_units_per_second(*num_assets_weight_hint))]
+		pub fn set_asset_units_per_second(
+			origin: OriginFor<T>,
+			asset_type: T::ForeignAssetType,
+			units_per_second: u128,
+			num_assets_weight_hint: u32,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			let asset_id: T::AssetId =
+				AssetTypeId::<T>::get(&asset_type).ok_or(Error::<T>::AssetTypeDoesNotExist)?;
+
+			// Grab supported assets
+			let mut supported_assets = SupportedFeePaymentAssets::<T>::get();
+
+			ensure!(
+				num_assets_weight_hint >= (supported_assets.len() as u32),
+				Error::<T>::TooLowNumAssetsWeightHint
+			);
+
+			// Only if the asset is not supported we need to push it
+			if let Err(index) = supported_assets.binary_search(&asset_id) {
+				supported_assets.insert(index, asset_id.clone());
+				SupportedFeePaymentAssets::<T>::put(supported_assets);
+			}
+
+			AssetIdUnitsPerSecond::<T>::insert(&asset_id, &units_per_second);
+
+			Self::deposit_event(Event::UnitsPerSecondChanged { asset_id, units_per_second });
+			Ok(())
+		}
+
+		/// Add the xcm type mapping for a existing assetId, other assetType still exists if any.
+		/// TODO: Weight is no longer related to num_assets_weight_hint; changes here needed
+		#[pallet::weight(T::WeightInfo::add_asset_type())]
+		pub fn add_asset_type(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			new_asset_type: T::ForeignAssetType,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			ensure!(AssetIdType::<T>::get(&asset_id).is_some(), Error::<T>::AssetIdDoesNotExist);
+			ensure!(
+				AssetTypeId::<T>::get(&new_asset_type).is_none(),
+				Error::<T>::AssetAlreadyExists
+			);
+
+			// Insert new asset type info
+			AssetIdType::<T>::insert(&asset_id, &new_asset_type);
+			AssetTypeId::<T>::insert(&new_asset_type, &asset_id);
+
+			// Remove previous asset type info
+			// AssetTypeId::<T>::remove(&previous_asset_type);
+
+			Self::deposit_event(Event::ForeignAssetRegistered {
+				asset_id,
+				asset_type: new_asset_type,
+			});
+			Ok(())
+		}
+
+		/// TODO: Weight is no longer related to num_assets_weight_hint; changes here needed
+		/// We do not allow the destroy of asset id so far
+		/// But we will remove all existing related fee settings. So will need re-implement by
+		/// extrinsic call. New default asset type is needed for the assignment of asset_id's
+		/// default asset type mapping
+		#[pallet::weight(T::WeightInfo::remove_asset_type(*num_assets_weight_hint))]
+		#[transactional]
+		pub fn remove_asset_type(
```

Comment:
TBH I'm not a fan of the signature of this function. 

As the user of this fn, I wouldn't expect that I have to pass in again the current asset_type and also the "new default" asset_type. I would expect that I provide the asset_id and the extrinsic will remove it from the storage.

Or if it just resets to a "default" asset type, we better call it something like `reset_asset_type`

Reply:
If you are not remove the default asset_type, you can pass nothing about the new default.
If you are removing the default asset_type, it will trigger the DefaultAssetTypeRemoved error.
I feel ok about the procedure here. Let's discuss this in a meeting.

Reply:
I have the same feeling as Kailai. The new_default_asset_type is a bit confusing to me.